README.md
# COVID-19-NL

Collect and update the information of positively tested cases for the Coronavirus in the Netherlands.

The data is updated automatically and stored in _data_ folder.

## Data sources

[Coronavirus (COVID-19) messages](https://www.volksgezondheidenzorg.info/onderwerp/infectieziekten/regionaal-internationaal/coronavirus-covid-19#node-coronavirus-covid-19-meldingen) provided by RIVM

**(Does not work anymore)**  
The direct link to download the csv:
`https://www.volksgezondheidenzorg.info/sites/default/files/map/detail_data/klik_corona{day}{month}{year}_rec_0.csv`

## Visualizations

![Number of infected in different provinces](./imgs/num_all_province_0330.png)
![Number of infected(less than 200) in different provinces](./imgs/num_sub_province_0330.png)
![Top 10 number of infected cities](./imgs/num_top_cities_0330.png)

## Docker supported

Also, I add a _Dockerfile_ here, if you want to run the codes within a Docker container.
The detail of the container can be find [here](https://github.com/ZequnZ/py-docker) , which is another repo of mine :)
Just follow:

1. Build a Docker container with the _Dockerfile_: `docker build -t zequnz/py-docker:0.0.5 .`
2. Run the container and get the containerID, here you have different options:
   - Just want to run the script: `docker run -itd -v $(pwd)/data:/app/data zequnz/py-docker:0.0.5 python src/workflow.nl`
   - Also run the jupyter notebook to play with the codes: `docker run -itd -p 8990:8990 -v $(pwd):/app zequnz/py-docker:0.0.5`
3. (Optional) To start the notebook, you need to go into the container: `docker exec -it <containerID> bash`
4. (Optional) Start the notebook: `jupyter notebook --ip 0.0.0.0 --port 8990 --no-browser --allow-root`

## TODO

- [x] Process the data with [the information of the Dutch municipalities](https://www.cbs.nl/-/media/_excel/2020/03/gemeenten%20alfabetisch%202020.xlsx)
- [x] Save the data in the _data_ folder
- [x] Create a Github workflow to update the data automatically
- [x] Data aggregation
  - [x] City-based data
  - [x] Province-based data
- [x] Provide some visualizations
  - [x] Lineplot showing the number of infected in different provinces
  - [x] Lineplot showing the top 10 number of infected in different cities

## Author

- ZequnZ

## License

This project is licensed under the MIT License - see the LICENSE.md file for details

## Acknowledgement

Thanks [covid19-eu-data](https://github.com/covid19-eu-zh/covid19-eu-data) for inspiring me.

Thanks Koelink for providing great data source in your [repo](https://github.com/Koelink/COVID-19-NLDATA)


docs/changelog.md
# Changelog

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/)

## Current version [0.0.8] - 2020-04-01

### Changed

- Get_info function: Due the website is updated, I need to modify the function to make it works again
- Change the display images in Readme

## [0.0.7] - 2020-03-21

### Added

 - Functions to generate lineplots to show the number of infected of different provinces and cities

## [0.0.6] - 2020-03-14

### Changed

- Get_info function: Due the website is updated, I need to modify the function to make it works again
- Save_info function: Same reason

## [0.0.5] - 2020-03-14

### Added

- Aggregation functions to generate the info in a city-level and province-level
- Do the aggregation after getting the data from RIVM page

## [0.0.4] - 2020-03-13

### Added

- Function to get the csv info
- Preprocess the csv_str: discard the last column
- Function to process and save the data
- History data (2020-03-12) as the workflow was broken at that day

### Changed

- Workflow

## [0.0.3] - 2020-03-08

### Added

- Enrich the daily data with the Dutch municipalities information
  - Add a column 'Province' in the saved csv file
  - Add a row listing the sum of infected number
- Add the history data (2020-02-27 - 2020-03-05)

### Changed

- Changed all the saved csv files into new format

## [0.0.2] - 2020-03-07

### Added

- Github workflow to update the data automatically

## [0.0.1] - 2020-03-07

### Added

- Functions to obtain and save the data
- Dockerfile to run the scripts

### Changed

- Update the dependency packages
- Update Readme

## [0.0.0] - 2020-03-06

### Added

- Initialize the whole project
- PR template
- TODO list
- Changelog


src/utils.py
""" Utils function """
import os
import re
import datetime
import pandas as pd
import seaborn as sns
import matplotlib.pyplot as plt


def agg_covid19_info_city(path=f"./data"):
    """
    Aggregate the data based on each city
    """

    # Get all daily info csv files
    # and Extract the date of each file
    csv_dict = {}
    with os.scandir(path) as files:
        for f in files:
            if f.name.startswith("NL_2020") and f.name.endswith(".csv"):
                csv_dict[re.findall(r"[0-9]{8}", f.name)[0]] = f.name

    # We can see that newer info contains more cities as the virus keeps spreading
    # Therefore we need to sort the dictionary
    res = ""
    for key in sorted(csv_dict, reverse=True):
        if type(res) == str:
            res = pd.read_csv(f"./data/{csv_dict[key]}")
            # Only use the useful columns
            res = res[["City", "Number"]]
            # Rename the column with the date
            res.rename(
                columns={"Number": key}, inplace=True,
            )
        else:
            df = pd.read_csv(f"./data/{csv_dict[key]}")
            # Only use the useful columns
            df = df[["City", "Number"]]
            # Rename the column with the date
            df.rename(
                columns={"Number": key}, inplace=True,
            )
            # Merge the dataframe
            res = res.merge(df, on="City", how="left")
    # We hope to see the info from the past to now
    res = res.reindex(columns=["City"] + sorted(csv_dict))
    res = res.fillna(0)

    res.to_csv(f"./data/NL_COVID19_info_city.csv", index=False)
    print(f"Successfully update and aggregate the city level data")


def agg_covid19_info_province(path=f"./data"):
    """
    Aggregate the data based on each province
    """

    # Get all daily info csv files
    # and Extract the date of each file
    csv_dict = {}
    with os.scandir(path) as files:
        for f in files:
            if f.name.startswith("NL_2020") and f.name.endswith(".csv"):
                csv_dict[re.findall(r"[0-9]{8}", f.name)[0]] = f.name

    # We can see that newer info contains more cities as the virus keeps spreading
    # Therefore we need to sort the dictionary
    res = ""
    for key in sorted(csv_dict, reverse=True):
        if type(res) == str:
            res = pd.read_csv(f"./data/{csv_dict[key]}")
            # Only use the useful columns
            res = res[["Province", "Number"]]
            # Rename the column with the date
            res.rename(
                columns={"Number": key}, inplace=True,
            )
            # Aggregate the info to get the sum of each province
            res = res.groupby("Province").sum()
            res.reset_index(inplace=True)
        else:
            df = pd.read_csv(f"./data/{csv_dict[key]}")
            # Only use the useful columns
            df = df[["Province", "Number"]]
            # Rename the column with the date
            df.rename(
                columns={"Number": key}, inplace=True,
            )
            df = df.groupby("Province").sum()
            df.reset_index(inplace=True)

            # Merge the dataframe
            res = res.merge(df, on="Province", how="left")
    # We hope to see the info from the past to now
    res = res.reindex(columns=["Province"] + sorted(csv_dict))
    res = res.fillna(0)

    # Put the row "SUM" as the last row
    res = res.reindex(
        [col for col in res.index if col != res[res.Province == "SUM"].index]
        + [res[res.Province == "SUM"].index[0]]
    )

    res.to_csv(f"./data/NL_COVID19_info_province.csv", index=False)
    print(f"Successfully update and aggregate the province level data")


def viz_preprocessing(df_path):
    """
    Preprocess the aggregation csv into a good format for visualization
    """
    df = pd.read_csv(df_path)
    res = df.T
    res = res.rename(columns=res.iloc[0]).drop(res.index[0])
    res = res.astype("int64")
    res.reset_index(inplace=True)
    res["index"] = res["index"].apply(
        lambda x: "{}-{}-{}".format(x[0:4], x[4:6], x[6:])
    )
    res["index"] = pd.to_datetime(res["index"])
    return res


def gen_p_graphs(df_path, threshold=200):
    """
    Generate graphs to show the number of infected in different NL provinces
    """
    df = viz_preprocessing(df_path)

    cols = list(df.columns)[1:-1]
    colors = sns.color_palette("bright", len(cols))
    plt.figure(figsize=(16, 9))
    for i, c in enumerate(cols):
        sns.lineplot(x="index", y=c, data=df, label=c, linewidth=1.5, color=colors[i])

    plt.grid(color="grey", linestyle="--", linewidth=0.5, which="both")
    plt.xlabel("Date", fontsize=16)
    plt.ylabel("Number of infected", fontsize=16)
    cur_time = (datetime.datetime.now() + datetime.timedelta(hours=1)).strftime(
        "%Y-%m-%d %H:%M:%S"
    )
    plt.title(
        f"Number of infected cases in different provinces of NL\n updated at {cur_time} (UTC+1)",
        fontsize=20,
    )
    plt.savefig(f"./imgs/num_all_province.png")
    # plt.show()

    col2 = df.iloc[-1, 1:] < threshold
    col2 = list(col2[col2].index)
    colors = sns.color_palette("Set2", len(col2))
    plt.figure(figsize=(16, 9))
    for i, c in enumerate(col2):
        sns.lineplot(x="index", y=c, data=df, label=c, linewidth=3, color=colors[i])

    plt.grid(color="grey", linestyle="--", linewidth=0.5, which="both")
    plt.xlabel("Date", fontsize=16)
    plt.ylabel("Number of infected", fontsize=16)
    plt.title(
        f"Number of infected cases(less than {threshold}) in different provinces of NL\n updated at {cur_time} (UTC+1)",
        fontsize=20,
    )
    plt.ylim(bottom=0, top=threshold)
    plt.savefig(f"./imgs/num_sub_province.png")
    # plt.show()
    print(f"Successfully generate graphs about province.")


def gen_c_graphs(df_path, top=10):
    """
    Generate graphs to show the top number of infected in NL cities.
    """
    df = viz_preprocessing(df_path)
    c = df.iloc[-1, 1:]
    cols = list(c.sort_values(ascending=False).index)[1:1 + top]
    colors = sns.color_palette("bright", len(cols))
    plt.figure(figsize=(16, 9))
    for i, c in enumerate(cols):
        sns.lineplot(x="index", y=c, data=df, label=c, linewidth=2, color=colors[i])

    plt.grid(color="grey", linestyle="--", linewidth=0.5, which="both")
    plt.xlabel("Date", fontsize=16)
    plt.ylabel("Number of infected", fontsize=16)
    cur_time = (datetime.datetime.now() + datetime.timedelta(hours=1)).strftime(
        "%Y-%m-%d %H:%M:%S"
    )
    plt.title(
        f"Top {top} number of infected cases in cities of NL\n updated at {cur_time} (UTC+1)",
        fontsize=20,
    )
    plt.savefig(f"./imgs/num_top_cities.png")
    # plt.show()
    print(f"Successfully generate graphs about city.")


src/workflow_nl.py
"""Workflow to obtain and save the data"""

from get_info_nl import get_coronavirus_info_nl_v4, save_info_nl_v3, RIVM_INFO_PAGE
from utils import (
    agg_covid19_info_city,
    agg_covid19_info_province,
    gen_p_graphs,
    gen_c_graphs,
)

if __name__ == "__main__":

    # As the layout of the page is changed, the previous workflow
    # doesn't work anymore (Noticed at 2020-03-12)
    #################################################
    #################################################
    # csv_link, csv_update_date = get_coronavirus_info_nl(
    #     RIVM_INFO_PAGE
    #     + "/onderwerp/infectieziekten/regionaal-internationaal/coronavirus-covid-19#!node-coronavirus-covid-19-meldingen"
    # )

    # save_info_nl(csv_link, csv_update_date)
    #################################################
    #################################################

    csv_str, csv_update_date = get_coronavirus_info_nl_v4(
        RIVM_INFO_PAGE
        + "/onderwerp/infectieziekten/regionaal-internationaal/coronavirus-covid-19#!node-coronavirus-covid-19-meldingen"
    )
    save_info_nl_v3(csv_str, csv_update_date)

    agg_covid19_info_city()
    agg_covid19_info_province()

    gen_p_graphs(f"./data/NL_COVID19_info_province.csv")
    gen_c_graphs(f"./data/NL_COVID19_info_city.csv")


src/get_info_nl.py
"""Define functions"""

import re
import pandas as pd
import numpy as np
import json
from io import StringIO
from urllib.request import urlopen
from bs4 import BeautifulSoup as soup


RIVM_INFO_PAGE = "https://www.volksgezondheidenzorg.info"
DUTCH_MONTH = {
    "januari": "01",
    "februari": "02",
    "maart": "03",
    "april": "04",
    "mei": "05",
    "juni": "06",
    "juli": "07",
    "augustus": "08",
    "september": "09",
    "oktober": "10",
    "november": "11",
    "december": "12",
}
YEAR = "2020"


def get_coronavirus_info_nl(info_link):
    """
    Get the coronavirus information from RIVM website.
    @return:
    csv_link: str, The csv download link
    csv_update_date: str, The date at which csv is updated
    """
    coronavirus_info_page = urlopen(info_link).read().decode("utf-8")
    soup_pattern = soup(coronavirus_info_page, features="lxml")

    # Find out this pattern by reading the source code of the page
    csv_download_class = soup_pattern.find("a", {"class": "csv-export detail-data"})

    # Get the csv link
    csv_link = RIVM_INFO_PAGE + csv_download_class.get("href")

    # Get the update date from the link
    csv_update_date = re.search(r"([0-9]{8})", csv_link)[0]
    csv_update_date = csv_update_date[4:] + csv_update_date[2:4] + csv_update_date[0:2]

    # Get the current time to get this information
    # now = datetime.now()
    # current_time = now.strftime("%H:%M:%S")

    # Print the result
    print(f"csv link:{csv_link}")
    print(f"csv update date:{csv_update_date}")

    return csv_link, csv_update_date


def save_info_nl(csv_link, csv_update_date):
    """
    Preprocess the data and save it as a csv file.
    """

    info_str = urlopen(csv_link).read().decode("utf-8")

    # Load the string into dataframe
    info_df = pd.read_csv(StringIO(info_str), sep=";")

    # Preprocess the dataframe
    info_df.dropna(inplace=True)
    info_df.reset_index(drop=True, inplace=True)
    KEEP_COLs = ["id", "Gemeente", "Aantal"]
    info_df = info_df[KEEP_COLs]

    # Make it consistent with the info of Dutch municipalities
    info_df.rename(
        columns={"Aantal": "Number", "id": "City_code", "Gemeente": "City"},
        inplace=True,
    )

    # if this column is string then transform it into integer
    if info_df["City_code"].dtype != np.dtype("int64"):
        info_df["City_code"] = info_df["City_code"].apply(
            lambda x: x if "," not in x else x.split(",")[0]
        )

    # Add the Province column
    dutch_info = pd.read_csv("./data/Dutch_municipalities_2020.csv")
    KEEP = ["Province", "City"]
    dutch_info = dutch_info[KEEP]
    info_df = info_df.merge(dutch_info, on="City", how="inner")

    info_df = info_df.reindex(
        columns=[col for col in info_df.columns if col != "Number"] + ["Number"]
    )

    # Add a row listing the total number
    info_df = info_df.append(
        pd.DataFrame(
            [["", "SUM", "SUM", sum(info_df["Number"])]], columns=list(info_df.columns)
        )
    )
    info_df.reset_index(drop=True, inplace=True)

    # Save the csv
    info_df.to_csv(f"./data/NL_{csv_update_date}.csv", index=False)


# Updated at 2020-03-13


def get_coronavirus_info_nl_v2(info_link):
    """
    Get the coronavirus information from RIVM website version 2
    Due to the fact that the info page is updated
    @return:
    csv_str: str, The csv info string
    """

    coronavirus_info_page = urlopen(info_link).read().decode("utf-8")
    soup_pattern = soup(coronavirus_info_page, features="lxml")

    csv_info_class = soup_pattern.find("div", {"id": "csvData"})
    csv_raw_str = csv_info_class.string

    # There is one extra column in the string
    # The usage for now is unknown
    # Before understanding it, just discard it
    csv_str = ""
    for line in csv_raw_str.split("\n"):
        csv_str += ";".join(line.split(";")[0:3]) + "\n"
    return csv_str


def save_info_nl_v2(csv_str):
    """
    Preprocess the data and save it as a csv file version 2
    """

    info_df = pd.read_csv(StringIO(csv_str), sep=";")

    # Rename the df
    info_df.rename(
        columns={"Aantal": "Number", "Gemeente": "City", "Gemnr": "City_code"},
        inplace=True,
    )

    # Get the update date
    date_string = info_df[info_df["City_code"] == -2]["City"].values[0]
    time_info = date_string.split(" ")
    month = DUTCH_MONTH.get(time_info[2])
    day = time_info[1]
    csv_update_date = YEAR + month + day
    print(f"csv update date:{csv_update_date}")

    # Translate the row about the number of patient with missing postcode or living abroad
    # into English
    inx = info_df[info_df["City_code"] == -1].index
    info_df.loc[inx, "City"] = "missing postcode and abroad"
    info_df.loc[inx, "City_code"] = ""

    # Delete useless info
    info_df.dropna(axis=0, inplace=True)

    # Change the order of columns
    info_df = info_df.reindex(
        columns=["City_code"] + [col for col in info_df.columns if col != "City_code"]
    )

    # Add the Province column
    dutch_info = pd.read_csv("./data/Dutch_municipalities_2020.csv")
    KEEP = ["Province", "City_code"]
    dutch_info = dutch_info[KEEP]
    info_df = info_df.merge(dutch_info, on="City_code", how="left")

    # Fill up the column province for the aggreration
    inx = info_df[info_df["City"] == "missing postcode and abroad"].index
    info_df.loc[inx, "Province"] = "missing postcode and abroad"

    # Change the order of columns
    info_df = info_df.reindex(
        columns=[col for col in info_df.columns if col != "Number"] + ["Number"]
    )
    info_df = info_df.fillna("")

    # Add a row listing the total number
    info_df = info_df.append(
        pd.DataFrame(
            [["", "SUM", "SUM", sum(info_df["Number"])]], columns=list(info_df.columns)
        )
    )
    info_df.reset_index(drop=True, inplace=True)

    # Save the csv
    info_df.to_csv(f"./data/NL_{csv_update_date}.csv", index=False)


# Updated at 2020-03-17


def get_coronavirus_info_nl_v3(info_link):
    """
    Get the coronavirus information from RIVM website version 3
    Due to the fact that the info page is updated
    @return:
    csv_str: str, The csv info string
    """

    coronavirus_info_page = urlopen(info_link).read().decode("utf-8")
    soup_pattern = soup(coronavirus_info_page, features="lxml")

    date_string = soup_pattern.find("p").string
    date_string = date_string.replace("\xa0", " ")
    time_info = date_string.split(" ")
    year = time_info[4]
    month = DUTCH_MONTH.get(time_info[3])
    day = time_info[2]
    csv_update_date = year + month + day

    csv_info_class = soup_pattern.find("div", {"id": "csvData"})
    csv_str = csv_info_class.string
    print(f"csv update date:{csv_update_date}")

    return csv_str, csv_update_date


def save_info_nl_v3(csv_str, csv_update_date):
    """
    Preprocess the data and save it as a csv file version 3
    The info from the website is updated at 2020-03-16, which contains 5 columns.
    For now I just keep the format consistent.
    I will think about to change it later.
    """

    KEEP_COL = ["Gemnr", "Gemeente", "Aantal"]
    info_df = pd.read_csv(StringIO(csv_str), sep=";")
    info_df = info_df[KEEP_COL]

    # Rename the df
    info_df.rename(
        columns={"Aantal": "Number", "Gemeente": "City", "Gemnr": "City_code"},
        inplace=True,
    )

    # Translate the row about the number of patient with missing postcode or living abroad
    # into English
    inx = info_df[info_df["City_code"] == -1].index
    info_df.loc[inx, "City"] = "missing postcode and abroad"
    info_df.loc[inx, "City_code"] = ""

    # Delete useless info
    info_df.dropna(axis=0, inplace=True)

    # Change the order of columns
    info_df = info_df.reindex(
        columns=["City_code"] + [col for col in info_df.columns if col != "City_code"]
    )

    # Add the Province column
    dutch_info = pd.read_csv("./data/Dutch_municipalities_2020.csv")
    KEEP = ["Province", "City_code"]
    dutch_info = dutch_info[KEEP]
    info_df = info_df.merge(dutch_info, on="City_code", how="left")

    # Fill up the column province for the aggreration
    inx = info_df[info_df["City"] == "missing postcode and abroad"].index
    info_df.loc[inx, "Province"] = "missing postcode and abroad"

    # Change the order of columns
    info_df = info_df.reindex(
        columns=[col for col in info_df.columns if col != "Number"] + ["Number"]
    )
    info_df = info_df.fillna("")

    # Add a row listing the total number
    info_df = info_df.append(
        pd.DataFrame(
            [["", "SUM", "SUM", sum(info_df["Number"])]], columns=list(info_df.columns)
        )
    )
    info_df.reset_index(drop=True, inplace=True)

    # Save the csv
    info_df.to_csv(f"./data/NL_{csv_update_date}.csv", index=False)


def get_coronavirus_info_nl_v4(info_link):
    """
    Get the coronavirus information from RIVM website version 3
    Due to the fact that the info page is updated
    @return:
    csv_str: str, The csv info string
    """

    coronavirus_info_page = urlopen(info_link).read().decode("utf-8")
    soup_pattern = soup(coronavirus_info_page, features="lxml")

    date_link = soup_pattern.find("div", {"id": "mapTitles"}).string
    decoder = json.JSONDecoder()
    date_string = decoder.decode(s=date_link)
    date_string = date_string["nl"]["mapSubtitle"]
    #     print(date_string)

    time_info = date_string.split(" ")[-1]
    day, month, year = time_info.split("-")
    month = month if len(month) != 1 else "0" + month
    day = day if len(day) != 1 else "0" + day
    csv_update_date = year + month + day

    csv_info_class = soup_pattern.find("div", {"id": "csvData"})
    csv_str = csv_info_class.string
    print(f"csv update date:{csv_update_date}")

    return csv_str, csv_update_date
